linkerd/app/core/src/trace.rs

```diff
@@ -33,25 +45,52 @@ pub fn init_log_compat() -> Result<(), Error> {
     tracing_log::LogTracer::init().map_err(Error::from)
 }
 
-pub fn with_filter(filter: impl AsRef<str>) -> (Dispatch, LevelHandle) {
+pub fn with_filter_and_format(
+    filter: impl AsRef<str>,
+    format: impl AsRef<str>,
+) -> (Dispatch, LevelHandle) {
     let filter = filter.as_ref();
+    let format = format.as_ref();
 
     // Set up the subscriber
     let start_time = clock::now();
-    let builder = FmtSubscriber::builder()
-        .with_timer(Uptime { start_time })
-        .with_env_filter(filter)
-        .with_filter_reloading()
-        .with_ansi(cfg!(test));
-    let handle = LevelHandle {
-        inner: builder.reload_handle(),
-    };
-    let dispatch = Dispatch::new(builder.finish());
-
-    (dispatch, handle)
+
+    match format {
+        "JSON" => {
+            let builder = FmtSubscriber::builder()
+                .json()
```

Comment:
I think that if we call the `json` method on the builder _last_, we can factor out most of the shared configuration:

```rust
let builder = FmtSubscriber::builder()
    .with_timer(Uptime { start_time })
    .with_filter(filter)
    .with_filter_reloading();

match format {
    "JSON" => {
        let builder = builder.json();
        // ...
    }
    // ...
}
```

Reply:
For whatever reason, we need to put json();  either first, or at least not last, if not the following error occurs:

```rust
pub fn reload_handle(&self) -> crate::reload::Handle<crate::EnvFilter, Formatter<N, E, W>>
Returns a Handle that may be used to reload the constructed subscriber's filter.

no method named `reload_handle` found for struct `tracing_subscriber::fmt::SubscriberBuilder<tracing_subscriber::fmt::format::json::JsonFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::json::Json, trace::Uptime>, tracing_subscriber::reload::Layer<tracing_subscriber::filter::env::EnvFilter, tracing_subscriber::layer::Layered<tracing_subscriber::fmt::fmt_layer::Layer<tracing_subscriber::registry::sharded::Registry, tracing_subscriber::fmt::format::DefaultFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::Full, trace::Uptime>>, tracing_subscriber::registry::sharded::Registry>>>` in the current scope

method not found in `tracing_subscriber::fmt::SubscriberBuilder<tracing_subscriber::fmt::format::json::JsonFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::json::Json, trace::Uptime>, tracing_subscriber::reload::Layer<tracing_subscriber::filter::env::EnvFilter, tracing_subscriber::layer::Layered<tracing_subscriber::fmt::fmt_layer::Layer<tracing_subscriber::registry::sharded::Registry, tracing_subscriber::fmt::format::DefaultFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::Full, trace::Uptime>>, tracing_subscriber::registry::sharded::Registry>>>`rustc(E0599)
trace.rs(64, 32): method not found in `tracing_subscriber::fmt::SubscriberBuilder<tracing_subscriber::fmt::format::json::JsonFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::json::Json, trace::Uptime>, tracing_subscriber::reload::Layer<tracing_subscriber::filter::env::EnvFilter, tracing_subscriber::layer::Layered<tracing_subscriber::fmt::fmt_layer::Layer<tracing_subscriber::registry::sharded::Registry, tracing_subscriber::fmt::format::DefaultFields, tracing_subscriber::fmt::format::Format<tracing_subscriber::fmt::format::Full, trace::Uptime>>, tracing_subscriber::registry::sharded::Registry>>>`
Peek Problem (⌥F8)
No quick fixes available
```

Reply:
Oh, I think `with_filter_reloading` needs to be called after `json`, because `json` changes the type of the formatter. So, you could still factor out the `with_timer` and `with_filter` calls, if you like (but it's not a big deal regardless).

Reply:
Perfect, thanks.